upup/pkg/fi/cloudup/gce/tpm/gcetpmverifier/tpmverifier.go

```diff
@@ -0,0 +1,241 @@
+/*
+Copyright 2021 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+package gcetpmverifier
+
+import (
+	"context"
+	"crypto"
+	"crypto/rsa"
+	"crypto/sha256"
+	"crypto/x509"
+	"encoding/json"
+	"encoding/pem"
+	"errors"
+	"fmt"
+	"net/http"
+	"strings"
+
+	compute "google.golang.org/api/compute/v1"
+	"google.golang.org/api/googleapi"
+	"k8s.io/kops/pkg/apis/nodeup"
+	"k8s.io/kops/pkg/bootstrap"
+	"k8s.io/kops/pkg/nodeidentity/gce"
+	"k8s.io/kops/upup/pkg/fi"
+	tpm "k8s.io/kops/upup/pkg/fi/cloudup/gce/tpm"
+)
+
+type tpmVerifier struct {
+	opt tpm.TPMVerifierOptions
+
+	computeClient *compute.Service
+}
+
+// NewTPMVerifier constructs a new TPM verifier for GCE.
+func NewTPMVerifier(opt *tpm.TPMVerifierOptions) (bootstrap.Verifier, error) {
+	ctx := context.Background()
+
+	computeClient, err := compute.NewService(ctx)
+	if err != nil {
+		return nil, fmt.Errorf("error building compute API client: %w", err)
+	}
+
+	return &tpmVerifier{
+		opt:           *opt,
+		computeClient: computeClient,
+	}, nil
+}
+
+var _ bootstrap.Verifier = &tpmVerifier{}
+
+func (v *tpmVerifier) Verify(ctx context.Context, token string, body []byte) (*bootstrap.VerifyResult, error) {
+	bootstrapRequest := &nodeup.BootstrapRequest{}
+	if err := json.Unmarshal(body, bootstrapRequest); err != nil {
+		return nil, fmt.Errorf("failed to decode request: %w", err)
+	}
+
+	claim := bootstrapRequest.Claim
+	if claim == nil {
+		return nil, fmt.Errorf("claim is required")
+	}
+
+	node := claim.Node
+	if claim.Node == nil {
+		return nil, fmt.Errorf("node is required")
+	}
+
+	if node.GCPProjectID == "" {
+		return nil, fmt.Errorf("gcpProjectID is required")
+	}
+	if node.Zone == "" {
+		return nil, fmt.Errorf("zone is required")
+	}
+	if node.Instance == "" {
+		return nil, fmt.Errorf("instance is required")
+	}
+
+	if node.GCPProjectID != v.opt.ProjectID {
+		return nil, fmt.Errorf("projectID does not match expected: got %q, want %q", node.GCPProjectID, v.opt.ProjectID)
```

Comment:
If the client's project ID has to match the server's project ID, why bother having the client pass a project ID in the first place?

Reply:
It's something we can relax in future.